### ISSUE TITLE
Enable basic type checking

### ISSUE THREAD

**Issue by qartik** [PECOS-packages/PECOS#40]
With my PR https://github.com/PECOS-packages/PECOS/pull/37 I presumed there was some basic type checking in place in the `development` branch, turns out I was mistaken.

It will be nice to enable it using `mypy` starting with the default config.

I'm happy to work on it in spare time.

Have a WIP PR at #41.

**Comment by qciaran** [comment]
Yeah, I would like to eventually to start adding `mypy` and stricter types. I would be curious to learn how these types are being used/affect pytket-phir. 

FYI, I am considering changing some of the typing for the inputs to `HybridEngine` to use classes that inherit from `typing.Protocol` because all I really want is duck typing there and not to enforce an ABC. However, I am not super familiar with it, but it seems to fit my intent more (to allow users to supply anything that supports a certain interface).

**Comment by qciaran** [comment]
Also, FYI, the Generic* classes are not meant to serve as base types but to be a generic example of the ABCs.
